### PR TITLE
Fix bug in hex_to_rgb, minor API changes, docstrings

### PR DESCRIPTION
**Pull request by Starwort**
## Motivation

Documentation strings and type-hinting are largely missing/inaccurate from the library; this pull request is the beginning of my attempt to fix that. One bug has been fixed and no major API changes have been made.

## Overview / Explanation for Changes

This pull request adds docstrings for several functions, changes some API quirks that are unlikely to be relied upon (such as the behaviour of `color_gradient(length_of_output=0)`, and fixes a bug in `hex_to_rgb` which made it fail for length-3 hex strings (e.g. `#fff`). It also changes the behaviour of `hex_to_rgb` when passed an invalid colour; it now throws a meaningful error instead of returning white under a bare `except`

## Oneline Summary of Changes

```
- Fixed bug where `hex_to_rgb` would fail on length-3 hexadecimal colour strings, even though it was supposed to work
- Added typehints
```

I'm sorry, I know this is two lines; but they're completely unrelated

## Testing Status

No tests needed; no major API changes

## Further Comments

This pull request is still a work-in-progress; I just felt that merging the bugfix would be worthwhile.

Please cherry-pick the bugfix and leave the pull request open until my documentation is complete

## Acknowledgements
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)

## Reviewer Checklist
- [ ] Newly added functions/classes are either private or have a docstring
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings
- [ ] The oneline summary has been included [in the wiki](https://github.com/ManimCommunity/manim/wiki/Changelog-for-next-release)

### CODE REVIEW

File: manim/utils/bezier.py [hunk 1]
```diff
@@ -57,11 +57,29 @@ def partial_bezier_points(points: np.ndarray, a: float, b: float) -> np.ndarray:
...
+
+    :param start: The start of the interpolation range
+    :type start: int | numpy.ndarray[numpy.float64]
+    :param end: The end of the interpolation range
+    :type end: int | numpy.ndarray[numpy.float64]
+    :param alpha: The position to interpolate, in the range [0, 1]
+    :type alpha: float
+    :return: The interpolated value
+    :rtype: float | numpy.ndarray[numpy.float64]
+    """
```
Comment (naveen521kk) [inline manim/utils/bezier.py:76]: Use NumPy style... 
https://github.com/ManimCommunity/manim/wiki/Documentation-guidelines-(WIP)

**Review by naveen521kk — CHANGES_REQUESTED** [review]
Thanks for your contribution.
Manim uses NumPy styled documentation, or rather docstrings. So, it needs to be formatted accordingly. 
See https://github.com/ManimCommunity/manim/wiki/Documentation-guidelines-(WIP) for more information.

File: tests/utils/GraphicalUnitTester.py [hunk 2]
```diff
@@ -135,7 +135,9 @@ def test(self, show_diff=False):
             if show_diff:
                 self._show_diff_helper(frame_data, expected_frame_data)
             assert test_result, (
-                f"The frames don't match. {str(self.scene).replace('Test', '')} has been modified."
+                f"The frames don't match. {str(self.scene).replace('Test', '')} has"
+                " been modified."
```
Comment (naveen521kk) [inline tests/utils/GraphicalUnitTester.py:139]: This modification seems irrelevant to me. We have black here to do that work and we are comfortable with line length around 88. Please revert.

File: tests/utils/GraphicalUnitTester.py [hunk 3]
```diff
@@ -135,7 +135,9 @@ def test(self, show_diff=False):
...
                 + "\nPlease ignore if it was intended."
-                + f"\nFirst unmatched index is at {first_incorrect_index}: {first_incorrect_point} != {expected_point}"
+                + f"\nFirst unmatched index is at {first_incorrect_index}:"
+                f" {first_incorrect_point} != {expected_point}"
```
Comment (naveen521kk) [inline tests/utils/GraphicalUnitTester.py:142]: same as previous

File: manim/utils/color.py [hunk 4]
```diff
@@ -84,11 +84,13 @@
     "ORANGE",
 ]
 
-from enum import Enum
 import random
+import re
```
Comment (naveen521kk) [inline manim/utils/color.py:88]: looks like an unused import...

File: manim/__main__.py [hunk 5]
```diff
@@ -19,8 +19,8 @@
 
 
 def open_file_if_needed(file_writer):
+    curr_stdout = sys.stdout
     if config["verbosity"] != "DEBUG":
```
Comment (naveen521kk) [inline manim/__main__.py:23]: Can you explain what does it change doing this here?

**Review by naveen521kk — CHANGES_REQUESTED** [review]
Well did you use any other automatic formatting tools other than black?

File: manim/utils/bezier.py (hunk 1 above)
Comment (Starwort) [reply manim/utils/bezier.py:76]: I can fix this; I was following a different docstring in this style; I did suggest just cherry-picking the bugfix, however :)

File: tests/utils/GraphicalUnitTester.py (hunk 2 above)
Comment (Starwort) [reply tests/utils/GraphicalUnitTester.py:139]: This change was performed by black.

File: tests/utils/GraphicalUnitTester.py (hunk 3 above)
Comment (Starwort) [reply tests/utils/GraphicalUnitTester.py:142]: As was this one

File: manim/__main__.py (hunk 5 above)
Comment (Starwort) [reply manim/__main__.py:23]: As explained in the commit, rebinding this name here has essentially no performance impact and reduces the number of warnings thrown by pylance

File: manim/utils/color.py (hunk 4 above)
Comment (Starwort) [reply manim/utils/color.py:88]: I use it in hex_to_rgb to enforce sensible behaviour for hex strings

File: manim/utils/color.py [hunk 6]
```diff
@@ -257,58 +259,188 @@ class Colors(Enum):
...
+    """
+    Convert `hex_code` to an RGB array
+
+    :param hex_code: Hexadecimal colour string
+    :type hex_code: str
+    :raises ValueError: `hex_code` is not a valid hexadecimal colour
+    :return: RGB array
+    :rtype: numpy.ndarray[numpy.float64]
+    """
+    if not re.fullmatch(r"#([0-9a-f]{3}|[0-9a-f]{6})", hex_code, flags=re.I):
```
Comment (Starwort) [inline manim/utils/color.py:370]: Here is where re is used :)

**Comment by Starwort** [comment]
> Well did you use any other automatic formatting tools other than black?

No, all formatting changes were performed by black

File: tests/utils/GraphicalUnitTester.py (hunk 2 above)
Comment (naveen521kk) [reply tests/utils/GraphicalUnitTester.py:139]: You should run black from the root of the folder. Only then black will the configurations. Revert this, and all other things because of this.

File: manim/utils/bezier.py (hunk 1 above)
Comment (naveen521kk) [reply manim/utils/bezier.py:76]: Docs looks good just follow numpy style :)

File: manim/__main__.py (hunk 5 above)
Comment (naveen521kk) [reply manim/__main__.py:23]: What warning did it throws?

File: tests/utils/GraphicalUnitTester.py (hunk 2 above)
Comment (Starwort) [reply tests/utils/GraphicalUnitTester.py:139]: Your `pyproject.toml` doesn't contain any black configuration? Black will behave in the same way...

File: manim/__main__.py (hunk 5 above)
Comment (Starwort) [reply manim/__main__.py:23]: 'Possibly Unbound Here'
As a human I can see that is isn't (as `config['verbosity']` should remain constant throughout the block) but there's no performance impact to this fix

File: tests/utils/GraphicalUnitTester.py (hunk 2 above)
Comment (naveen521kk) [reply tests/utils/GraphicalUnitTester.py:139]: [Black defaults to line length 88.](https://black.readthedocs.io/en/stable/the_black_code_style.html#line-length) Then why does it break this? Are you sure you using black? Or does black uses any personal configuration?

File: manim/utils/caching.py [hunk 7]
```diff
@@ -37,7 +38,11 @@ def wrapper(self, scene, *args, **kwargs):
             )
             if self.file_writer.is_already_cached(hash_play):
                 logger.info(
-                    f"Animation {self.num_plays} : Using cached data (hash : %(hash_play)s)",
+                    f"Animation {self.num_plays} : "
+                    "Using cached data (hash : %(hash_play)s)",
+                    # an f-string is being used here; couldn't
+                    # the format dictionary be replaced with a
+                    # second interpolation?
```
Comment (huguesdevimeux) [inline manim/utils/caching.py:45]: That's a fair question, and the answer is no. 

This is for logs tests of the librairy, as hashes are inconsitent over the runs we want to avoid having them in the logs. 
This is part of the trick to remove them :)

File: tests/utils/GraphicalUnitTester.py (hunk 2 above)
Comment (Starwort) [reply tests/utils/GraphicalUnitTester.py:139]: That line is originally 99c long. Black has wrapped it to fit within 88 characters

Comment (naveen521kk) [reply tests/utils/GraphicalUnitTester.py:139]: well now I think the test's folder isn't tested to be formatted I think...

**Comment by behackl** [comment]
@naveen521kk, have your change requests here been resolved? Is there anything left to do for this PR?

**Comment by naveen521kk** [comment]
The docstrings aren't in Numpy format, it needs to be changed.

**Comment by huguesdevimeux** [comment]
Any update on this ?

**Comment by behackl** [comment]
Closing due to inactivity. Please feel free to reopen this PR / comment here again as soon as you get around to make the requested changes.
